File: backend/routes/packages.py
"""
TravelVista — Packages Routes
"""
import json
import time
from datetime import date
from flask import Blueprint, request, jsonify
from flask_jwt_extended import jwt_required, get_jwt_identity
from models import get_db, package_to_dict

packages_bp = Blueprint('packages', __name__, url_prefix='/api/packages')


def _is_admin(uid):
    conn = get_db()
    row = conn.execute("SELECT role FROM users WHERE id = ?", (uid,)).fetchone()
    conn.close()
    return row and row['role'] == 'admin'


@packages_bp.route('', methods=['GET'])
def list_packages():
    """Public: list active packages with optional search/filter/sort."""
    q = request.args.get('q', '').strip().lower()
    category = request.args.get('category', '')
    min_price = request.args.get('minPrice', type=float)
    max_price = request.args.get('maxPrice', type=float)
    sort_by = request.args.get('sortBy', '')

    conn = get_db()
    rows = conn.execute("SELECT * FROM packages WHERE active = 1").fetchall()
    conn.close()

    result = [package_to_dict(r) for r in rows]

    if q:
        result = [p for p in result if
                  q in p['title'].lower() or
                  q in p['location'].lower() or
                  q in p['category'].lower()]

    if category:
        result = [p for p in result if p['category'] == category]
    if min_price is not None:
        result = [p for p in result if p['price'] >= min_price]
    if max_price is not None:
        result = [p for p in result if p['price'] <= max_price]

    if sort_by == 'price-low':
        result.sort(key=lambda p: p['price'])
    elif sort_by == 'price-high':
        result.sort(key=lambda p: p['price'], reverse=True)
    elif sort_by == 'rating':
        result.sort(key=lambda p: p['rating'], reverse=True)
    elif sort_by == 'popular':
        result.sort(key=lambda p: p['reviewCount'], reverse=True)

    return jsonify(result)


@packages_bp.route('/all', methods=['GET'])
@jwt_required()
def list_all():
    """Admin: list ALL packages including inactive."""
    uid = get_jwt_identity()
    if not _is_admin(uid):
        return jsonify({'message': 'Forbidden'}), 403

    conn = get_db()
    rows = conn.execute("SELECT * FROM packages ORDER BY created_at DESC").fetchall()
    conn.close()
    return jsonify([package_to_dict(r) for r in rows])


@packages_bp.route('/<pkg_id>', methods=['GET'])
def get_package(pkg_id):
    conn = get_db()
    row = conn.execute("SELECT * FROM packages WHERE id = ?", (pkg_id,)).fetchone()
    conn.close()
    if not row:
        return jsonify({'message': 'Package not found'}), 404
    return jsonify(package_to_dict(row))


@packages_bp.route('', methods=['POST'])
@jwt_required()
def create_package():
    uid = get_jwt_identity()
    if not _is_admin(uid):
        return jsonify({'message': 'Forbidden'}), 403

    data = request.get_json() or {}
    pkg_id = 'pkg-' + str(int(time.time() * 1000))

    conn = get_db()
    conn.execute(
        """INSERT INTO packages
           (id,title,location,duration,price,original_price,max_persons,rating,review_count,
            image,gallery,category,description,highlights,itinerary,inclusions,exclusions,
            active,featured,created_at)
           VALUES (?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?)""",
        (
            pkg_id,
            data.get('title', ''),
            data.get('location', ''),
            data.get('duration', ''),
            data.get('price', 0),
            data.get('originalPrice', 0),
            data.get('maxPersons', 20),
            0, 0,
            data.get('image', ''),
            json.dumps(data.get('gallery', [])),
            data.get('category', ''),
            data.get('description', ''),
            json.dumps(data.get('highlights', [])),
            json.dumps(data.get('itinerary', [])),
            json.dumps(data.get('inclusions', [])),
            json.dumps(data.get('exclusions', [])),
            1 if data.get('active', True) else 0,
            1 if data.get('featured', False) else 0,
            date.today().isoformat(),
        )
    )
    conn.commit()
    row = conn.execute("SELECT * FROM packages WHERE id = ?", (pkg_id,)).fetchone()
    conn.close()
    return jsonify(package_to_dict(row)), 201


@packages_bp.route('/<pkg_id>', methods=['PUT'])
@jwt_required()
def update_package(pkg_id):
    uid = get_jwt_identity()
    if not _is_admin(uid):
        return jsonify({'message': 'Forbidden'}), 403

    data = request.get_json() or {}
    conn = get_db()

    existing = conn.execute("SELECT * FROM packages WHERE id = ?", (pkg_id,)).fetchone()
    if not existing:
        conn.close()
        return jsonify({'message': 'Package not found'}), 404

    conn.execute(
        """UPDATE packages SET
           title=?, location=?, duration=?, price=?, original_price=?, max_persons=?,
           image=?, gallery=?, category=?, description=?, highlights=?,
           itinerary=?, inclusions=?, exclusions=?, active=?, featured=?
           WHERE id=?""",
        (
            data.get('title', existing['title']),
            data.get('location', existing['location']),
            data.get('duration', existing['duration']),
            data.get('price', existing['price']),
            data.get('originalPrice', existing['original_price']),
            data.get('maxPersons', existing['max_persons']),
            data.get('image', existing['image']),
            json.dumps(data.get('gallery', json.loads(existing['gallery'] or '[]'))),
            data.get('category', existing['category']),
            data.get('description', existing['description']),
            json.dumps(data.get('highlights', json.loads(existing['highlights'] or '[]'))),
            json.dumps(data.get('itinerary', json.loads(existing['itinerary'] or '[]'))),
            json.dumps(data.get('inclusions', json.loads(existing['inclusions'] or '[]'))),
            json.dumps(data.get('exclusions', json.loads(existing['exclusions'] or '[]'))),
            1 if data.get('active', bool(existing['active'])) else 0,
            1 if data.get('featured', bool(existing['featured'])) else 0,
            pkg_id,
        )
    )
    conn.commit()
    row = conn.execute("SELECT * FROM packages WHERE id = ?", (pkg_id,)).fetchone()
    conn.close()
    return jsonify(package_to_dict(row))


@packages_bp.route('/<pkg_id>', methods=['DELETE'])
@jwt_required()
def delete_package(pkg_id):
    uid = get_jwt_identity()
    if not _is_admin(uid):
        return jsonify({'message': 'Forbidden'}), 403

    conn = get_db()
    conn.execute("DELETE FROM packages WHERE id = ?", (pkg_id,))
    conn.commit()
    conn.close()
    return jsonify({'success': True})


@packages_bp.route('/<pkg_id>/toggle', methods=['PATCH'])
@jwt_required()
def toggle_active(pkg_id):
    uid = get_jwt_identity()
    if not _is_admin(uid):
        return jsonify({'message': 'Forbidden'}), 403

    conn = get_db()
    row = conn.execute("SELECT * FROM packages WHERE id = ?", (pkg_id,)).fetchone()
    if not row:
        conn.close()
        return jsonify({'message': 'Package not found'}), 404

    new_active = 0 if row['active'] else 1
    conn.execute("UPDATE packages SET active = ? WHERE id = ?", (new_active, pkg_id))
    conn.commit()
    row = conn.execute("SELECT * FROM packages WHERE id = ?", (pkg_id,)).fetchone()
    conn.close()
    return jsonify(package_to_dict(row))

File: api/packages.php
<?php
/**
 * TravelVista — Packages API
 * GET    (no action)        — List active packages (public)
 * GET    ?action=all        — List ALL packages (admin)
 * GET    ?id=X              — Get single package
 * POST                      — Create package (admin)
 * PUT    ?id=X              — Update package (admin)
 * DELETE ?id=X              — Delete package (admin)
 * PATCH  ?id=X&action=toggle — Toggle active (admin)
 */

header('Content-Type: application/json');
header('Access-Control-Allow-Origin: *');
header('Access-Control-Allow-Methods: GET, POST, PUT, DELETE, PATCH, OPTIONS');
header('Access-Control-Allow-Headers: Content-Type, Authorization');

if ($_SERVER['REQUEST_METHOD'] === 'OPTIONS') { http_response_code(200); exit; }

require_once __DIR__ . '/../config/jwt_helper.php';

$method = $_SERVER['REQUEST_METHOD'];
$action = $_GET['action'] ?? '';
$id = $_GET['id'] ?? '';
if (!$action && !$id) {
    $uri = $_SERVER['REQUEST_URI'] ?? '';
    if (preg_match('#/api/packages/(\d+)/toggle#', $uri, $m)) { $id = $m[1]; $action = 'toggle'; }
    elseif (preg_match('#/api/packages/all#', $uri)) { $action = 'all'; }
    elseif (preg_match('#/api/packages/(\d+)#', $uri, $m)) { $id = $m[1]; }
}

// ── GET: List active packages (public) ──
if ($method === 'GET' && !$action && !$id) {
    $q = strtolower(trim($_GET['q'] ?? ''));
    $category = $_GET['category'] ?? '';
    $minPrice = isset($_GET['minPrice']) ? (float)$_GET['minPrice'] : null;
    $maxPrice = isset($_GET['maxPrice']) ? (float)$_GET['maxPrice'] : null;
    $sortBy = $_GET['sortBy'] ?? '';

    $db = getDB();
    $rows = $db->query("SELECT * FROM packages WHERE active = 1")->fetchAll();
    $result = array_map('package_to_dict', $rows);

    if ($q) {
        $result = array_filter($result, function($p) use ($q) {
            return strpos(strtolower($p['title']), $q) !== false ||
                   strpos(strtolower($p['location']), $q) !== false ||
                   strpos(strtolower($p['category']), $q) !== false;
        });
    }
    if ($category) {
        $result = array_filter($result, fn($p) => $p['category'] === $category);
    }
    if ($minPrice !== null) {
        $result = array_filter($result, fn($p) => $p['price'] >= $minPrice);
    }
    if ($maxPrice !== null) {
        $result = array_filter($result, fn($p) => $p['price'] <= $maxPrice);
    }

    $result = array_values($result);

    if ($sortBy === 'price-low') usort($result, fn($a, $b) => $a['price'] <=> $b['price']);
    elseif ($sortBy === 'price-high') usort($result, fn($a, $b) => $b['price'] <=> $a['price']);
    elseif ($sortBy === 'rating') usort($result, fn($a, $b) => $b['rating'] <=> $a['rating']);
    elseif ($sortBy === 'popular') usort($result, fn($a, $b) => $b['reviewCount'] <=> $a['reviewCount']);

    json_response($result);
}

// ── GET: All packages (admin) ──
if ($method === 'GET' && $action === 'all') {
    require_admin();
    $db = getDB();
    $rows = $db->query("SELECT * FROM packages ORDER BY created_at DESC")->fetchAll();
    json_response(array_map('package_to_dict', $rows));
}

// ── GET: Single package ──
if ($method === 'GET' && $id) {
    $db = getDB();
    $stmt = $db->prepare("SELECT * FROM packages WHERE id = ?");
    $stmt->execute([$id]);
    $row = $stmt->fetch();
    if (!$row) json_response(['message' => 'Package not found'], 404);
    json_response(package_to_dict($row));
}

// ── POST: Create package ──
if ($method === 'POST') {
    require_admin();
    $data = get_json_body();
    $db = getDB();

    $stmt = $db->prepare(
        "INSERT INTO packages (title, location, duration, price, original_price, max_persons, rating, review_count,
         image, gallery, category, description, highlights, itinerary, inclusions, exclusions, active, featured, created_at)
         VALUES (?, ?, ?, ?, ?, ?, 0, 0, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)"
    );
    $stmt->execute([
        $data['title'] ?? '',
        $data['location'] ?? '',
        $data['duration'] ?? '',
        $data['price'] ?? 0,
        $data['originalPrice'] ?? 0,
        $data['maxPersons'] ?? 20,
        $data['image'] ?? '',
        json_encode($data['gallery'] ?? []),
        $data['category'] ?? '',
        $data['description'] ?? '',
        json_encode($data['highlights'] ?? []),
        json_encode($data['itinerary'] ?? []),
        json_encode($data['inclusions'] ?? []),
        json_encode($data['exclusions'] ?? []),
        ($data['active'] ?? true) ? 1 : 0,
        ($data['featured'] ?? false) ? 1 : 0,
        date('Y-m-d'),
    ]);

    $pkgId = $db->lastInsertId();
    $stmt = $db->prepare("SELECT * FROM packages WHERE id = ?");
    $stmt->execute([$pkgId]);
    json_response(package_to_dict($stmt->fetch()), 201);
}

// ── PUT: Update package ──
if ($method === 'PUT' && $id) {
    require_admin();
    $data = get_json_body();
    $db = getDB();

    $stmt = $db->prepare("SELECT * FROM packages WHERE id = ?");
    $stmt->execute([$id]);
    $existing = $stmt->fetch();
    if (!$existing) json_response(['message' => 'Package not found'], 404);

    $stmt = $db->prepare(
        "UPDATE packages SET title=?, location=?, duration=?, price=?, original_price=?, max_persons=?,
         image=?, gallery=?, category=?, description=?, highlights=?, itinerary=?, inclusions=?, exclusions=?,
         active=?, featured=? WHERE id=?"
    );
    $stmt->execute([
        $data['title'] ?? $existing['title'],
        $data['location'] ?? $existing['location'],
        $data['duration'] ?? $existing['duration'],
        $data['price'] ?? $existing['price'],
        $data['originalPrice'] ?? $existing['original_price'],
        $data['maxPersons'] ?? $existing['max_persons'],
        $data['image'] ?? $existing['image'],
        json_encode($data['gallery'] ?? json_decode($existing['gallery'] ?: '[]', true)),
        $data['category'] ?? $existing['category'],
        $data['description'] ?? $existing['description'],
        json_encode($data['highlights'] ?? json_decode($existing['highlights'] ?: '[]', true)),
        json_encode($data['itinerary'] ?? json_decode($existing['itinerary'] ?: '[]', true)),
        json_encode($data['inclusions'] ?? json_decode($existing['inclusions'] ?: '[]', true)),
        json_encode($data['exclusions'] ?? json_decode($existing['exclusions'] ?: '[]', true)),
        isset($data['active']) ? ($data['active'] ? 1 : 0) : $existing['active'],
        isset($data['featured']) ? ($data['featured'] ? 1 : 0) : $existing['featured'],
        $id,
    ]);

    $stmt = $db->prepare("SELECT * FROM packages WHERE id = ?");
    $stmt->execute([$id]);
    json_response(package_to_dict($stmt->fetch()));
}

// ── DELETE: Delete package ──
if ($method === 'DELETE' && $id) {
    require_admin();
    $db = getDB();
    $db->prepare("DELETE FROM packages WHERE id = ?")->execute([$id]);
    json_response(['success' => true]);
}

// ── PATCH: Toggle active ──
if ($method === 'PATCH' && $id && $action === 'toggle') {
    require_admin();
    $db = getDB();
    $stmt = $db->prepare("SELECT * FROM packages WHERE id = ?");
    $stmt->execute([$id]);
    $row = $stmt->fetch();
    if (!$row) json_response(['message' => 'Package not found'], 404);

    $newActive = $row['active'] ? 0 : 1;
    $db->prepare("UPDATE packages SET active = ? WHERE id = ?")->execute([$newActive, $id]);

    $stmt = $db->prepare("SELECT * FROM packages WHERE id = ?");
    $stmt->execute([$id]);
    json_response(package_to_dict($stmt->fetch()));
}

json_response(['message' => 'Invalid request'], 400);
?>
